### PR TITLE
Add validation when load page

### PR DESCRIPTION
**Pull request by SuZhou-Joe**
### Description

- Change the implementation of workspaceId storage from Cookie to querystring & sessionstorage.
- Expose currentWorkspaceId$ and workspaceList$ for subscription.
- Add landing page logic that handles with workspaceId(from querystring or sessionstorage).
- Fix router validation schema

### Check List

- [ ] All tests pass
  - [ ] `yarn test:jest`
  - [ ] `yarn test:jest_integration`
  - [ ] `yarn test:ftr`
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] Update [CHANGELOG.md](./../CHANGELOG.md)
- [x] Commits are signed per the DCO using --signoff

### CODE REVIEW

File: src/core/public/workspace/workspaces_client.ts [hunk 1]
```diff
@@ -38,33 +39,103 @@ type IResponse<T> =
...
+        result: null,
+      };
+    } else {
+      return workspaceResp;
+    }
   }
 
   public async exitWorkspace(): Promise<IResponse<null>> {
-    return this.http.post(this.getPath(['_exit']));
+    this.currentWorkspaceId$.next('');
```
Comment (ruanyl) [inline src/core/public/workspace/workspaces_client.ts:123]: Q: do we still need this function`exitWorkspace` if user always has a workspace?

File: src/core/public/workspace/workspaces_client.ts [hunk 2]
```diff
@@ -83,36 +154,54 @@ export class WorkspacesClient {
...
    * @returns
    */
-  public create = (
+  public async create(
     attributes: Omit<WorkspaceAttribute, 'id'>
-  ): Promise<IResponse<WorkspaceAttribute>> => {
+  ): Promise<IResponse<WorkspaceAttribute>> {
     if (!attributes) {
-      return Promise.reject(new Error('requires attributes'));
+      return {
```
Comment (ruanyl) [inline src/core/public/workspace/workspaces_client.ts:165]: Do we still need this validation here if server side has defined the validation schema?

File: src/core/public/workspace/workspaces_client.ts [hunk 3]
```diff
@@ -38,33 +39,103 @@ type IResponse<T> =
...
+          error: error.message,
+        } as T;
+      }
+
+      return {
+        success: false,
+        error: 'Unknown error',
+      } as T;
+    }
+  };
```
Comment (ruanyl) [inline src/core/public/workspace/workspaces_client.ts:93]:
```suggestion
  private catchedFetch = async <T = any>(
    path: string,
    options: HttpFetchOptions
  ): Promise<IResponse<T>> => {
    try {
      return await this.http.fetch<IResponse<T>>(path, options);
    } catch (error: unknown) {
      if (error instanceof HttpFetchError || error instanceof Error) {
        return {
          success: false,
          error: error.message,
        };
      }

      return {
        success: false,
        error: 'Unknown error',
      };
    }
  };
```

File: src/core/public/workspace/workspaces_client.ts [hunk 4]
```diff
@@ -83,36 +154,54 @@ export class WorkspacesClient {
...
    * Deletes a workspace
    *
    * @param id
    * @returns
    */
-  public delete = (id: string): Promise<IResponse<null>> => {
+  public async delete(id: string): Promise<IResponse<null>> {
     if (!id) {
-      return Promise.reject(new Error('requires id'));
+      return {
```
Comment (ruanyl) [inline src/core/public/workspace/workspaces_client.ts:195]: The same question here: shall we just let server side to handle the validation?

File: src/core/public/workspace/workspaces_client.ts [hunk 5]
```diff
@@ -167,19 +259,36 @@ export class WorkspacesClient {
...
    * @returns
    */
-  public update(id: string, attributes: Partial<WorkspaceAttribute>): Promise<IResponse<boolean>> {
+  public async update(
+    id: string,
+    attributes: Partial<WorkspaceAttribute>
+  ): Promise<IResponse<boolean>> {
     if (!id || !attributes) {
-      return Promise.reject(new Error('requires id and attributes'));
+      return {
```
Comment (ruanyl) [inline src/core/public/workspace/workspaces_client.ts:271]: The same question here

File: src/plugins/workspace/public/plugin.ts [hunk 6]
```diff
@@ -4,17 +4,79 @@
...
+        sessionStorage.setItem(WORKSPACE_ID_IN_SESSION_STORAGE, newWorkspaceId);
+      } catch (e) {
+        /**
+         * in incognize mode, this method may throw an error
+         * */
+      }
+    });
+  }
+  private getWorkpsaceIdFromQueryString(): string {
+    const querystringObject = parse(window.location.search.replace(/^\??/, ''));
```
Comment (ruanyl) [inline src/plugins/workspace/public/plugin.ts:32]: Perhaps we can use `new URLSearchParams(window.location.search)`

File: src/core/server/workspaces/routes/index.ts [hunk 7]
```diff
@@ -5,13 +5,11 @@
...
 import { Logger } from '../../logging';
-import { IWorkspaceDBImpl, WORKSPACES_API_BASE_URL, WORKSPACE_ID_COOKIE_NAME } from '../types';
+import { IWorkspaceDBImpl } from '../types';
 
-function getCookieValue(cookieString: string, cookieName: string): string | null {
-  const regex = new RegExp(`(?:(?:^|.*;\\s*)${cookieName}\\s*\\=\\s*([^;]*).*$)|^.*$`);
-  const match = cookieString.match(regex);
-  return match ? match[1] : null;
-}
+export const WORKSPACES_API_BASE_URL = '/api/workspaces';
```
Comment (ruanyl) [inline src/core/server/workspaces/routes/index.ts:10]: Seems no need to export, it's not used by other modules.

File: src/core/public/workspace/workspaces_client.ts (hunk 1 above)
Comment (SuZhou-Joe) [reply src/core/public/workspace/workspaces_client.ts:123]: "Whether user always has a workspace" is still under discussion, there is no harm that we keep this API for now I think.

Comment (ruanyl) [reply src/core/public/workspace/workspaces_client.ts:123]: Makes sense, thanks

File: src/core/public/workspace/workspaces_client.ts [hunk 8]
```diff
@@ -38,33 +39,107 @@ type IResponse<T> =
...
+      }
+    });
+
+    /**
+     * Initialize workspace list
+     */
+    this.updateWorkspaceListAndNotify();
   }
 
+  private catchedFetch = async <T extends IResponse<any>>(
```
Comment (ruanyl) [inline src/core/public/workspace/workspaces_client.ts:74]: Get a warning which says `catched is an unknown word`, maybe rename it?

**Pull request merged**